pkg/controllers/allocation/controller.go

```diff
@@ -79,66 +78,58 @@ func NewController(kubeClient client.Client, coreV1Client corev1.CoreV1Interface
 			return reconcile.Result{}, nil
 		}
-		return result.RetryIfError(ctx, err)
+		return reconcile.Result{}, err
 	}
-
-	// 2. Wait on a pod batch
+	// Wait on a pod batch
 	logging.FromContext(ctx).Infof("Waiting to batch additional pods")
 	c.Batcher.Wait(provisioner)
 
-	// 3. Filter pods
+	// Filter pods
 	pods, err := c.Filter.GetProvisionablePods(ctx, provisioner)
 	if err != nil {
-		return result.RetryIfError(ctx, fmt.Errorf("filtering pods, %w", err))
+		return reconcile.Result{}, fmt.Errorf("filtering pods, %w", err)
 	}
 	logging.FromContext(ctx).Infof("Found %d provisionable pods", len(pods))
 	if len(pods) == 0 {
+		logging.FromContext(ctx).Infof("Watching for pod events")
 		return reconcile.Result{}, nil
 	}
-	// 4. Group by constraints
+	// Group by constraints
 	schedules, err := c.Scheduler.Solve(ctx, provisioner, pods)
 	if err != nil {
-		return result.RetryIfError(ctx, fmt.Errorf("solving scheduling constraints, %w", err))
+		return reconcile.Result{}, fmt.Errorf("solving scheduling constraints, %w", err)
 	}
-	// 5. Get Instance Types Options
+	// Get Instance Types Options
 	instanceTypes, err := c.CloudProvider.GetInstanceTypes(ctx)
 	if err != nil {
-		return result.RetryIfError(ctx, fmt.Errorf("getting instance types, %w", err))
+		return reconcile.Result{}, fmt.Errorf("getting instance types, %w", err)
 	}
-
-	// 6. Binpack each group
-	packings := []*binpacking.Packing{}
-	for _, schedule := range schedules {
-		packings = append(packings, c.Packer.Pack(ctx, schedule, instanceTypes)...)
-	}
-
-	// 7. Create capacity
-	errs := make([]error, len(packings))
-	workqueue.ParallelizeUntil(ctx, len(packings), len(packings), func(index int) {
-		packing := packings[index]
-		errs[index] = <-c.CloudProvider.Create(ctx, packing.Constraints, packing.InstanceTypeOptions, func(node *v1.Node) error {
-			node.Labels = functional.UnionStringMaps(
-				node.Labels,
-				packing.Constraints.Labels,
-				map[string]string{v1alpha4.ProvisionerNameLabelKey: provisioner.Name},
-			)
-			node.Spec.Taints = append(node.Spec.Taints, packing.Constraints.Taints...)
-			return c.Binder.Bind(ctx, node, packing.Pods)
-		})
+	// Create capacity
+	var errs error
+	workqueue.ParallelizeUntil(ctx, len(schedules), len(schedules), func(index int) {
+		for _, packing := range c.Packer.Pack(ctx, schedules[index], instanceTypes) {
+			if err := <-c.CloudProvider.Create(ctx, packing.Constraints, packing.InstanceTypeOptions, func(node *v1.Node) error {
+				node.Labels = functional.UnionStringMaps(
+					node.Labels,
+					packing.Constraints.Labels,
+					map[string]string{v1alpha4.ProvisionerNameLabelKey: provisioner.Name},
+				)
+				node.Spec.Taints = append(node.Spec.Taints, packing.Constraints.Taints...)
+				return c.Binder.Bind(ctx, node, packing.Pods)
+			}); err != nil {
+				errs = multierr.Append(errs, err)
```

Comment:
Just double checking, is this concurrency-safe? The older code seemed more explicitly so (since each goroutine(?) would only update a specific index (caveat: IIUC)).

Reply:
You're absolutely correct.